src/common/dependency-manager/local-manager.ts

```diff
@@ -119,72 +82,69 @@ export default class LocalDependencyManager extends DependencyManager {
   /**
    * @override
    */
-  async loadDependencies(parent_node: ServiceNode, recursive = true) {
-    if (parent_node instanceof ExternalNode) { return; }
-
-    const dependency_resolvers = [];
-    for (const [dep_name, dep_id] of Object.entries(parent_node.service_config.getDependencies())) {
-      const env_services = this.environment.getServiceDetails(`${dep_name}:${dep_id}`);
-      if (env_services?.host && env_services?.port) {
-        await this.loadExternalService(env_services, `${dep_name}:${dep_id}`);
-      } else {
-        let dep_node: ServiceNode;
-        if (env_services?.debug?.path) {
-          const svc_path = path.join(path.dirname(this.config_path), env_services.debug.path);
-          dep_node = await this.loadLocalService(svc_path);
-        } else if (this.linked_services.hasOwnProperty(dep_name)) {
-          console.log(`Using locally linked ${chalk.blue(dep_name)} found at ${chalk.blue(this.linked_services[dep_name])}`);
-          dep_node = await this.loadLocalService(this.linked_services[dep_name]);
-        } else {
-          dep_node = await this.loadService(dep_name, dep_id);
-        }
-
-        this.graph.addNode(dep_node);
-        const edge = new ServiceEdge(parent_node.ref, dep_node.ref);
-        this.graph.addEdge(edge);
+  async loadService(service_name: string, service_tag: string, recursive = true): Promise<ServiceNode | ExternalNode> {
+    const ref = `${service_name}:${service_tag}`;
+    const existing_node = this.graph.nodes_map.get(ref);
+    if (existing_node) {
+      return existing_node as ServiceNode | ExternalNode;
```

Comment:
Why does this need a cast?

What scenarios might this not be a ServiceNode or an ExternalNode and can we protect against that?

Reply:
It will never not be a ServiceNode/ExternalNode unless another node has the same ref which would be bad, but nodes_map is a map of DependencyNode so it doesn't know better.